### ISSUE TITLE
Native Cosmic applications are slow to present window.

### ISSUE THREAD

**Issue by XV-02** [pop-os/cosmic-comp#348]
I believe this behaviour started in the week or so prior to 9cb1adb , but I can't be sure. 

A handful of Cosmic applications (Files, Term, Edit) are showing a significant and frustrating delay between when I launch them and when their windows appears. It looks like the processes underlying the applications are themselves processing quickly - the `cosmic-term` process seems to be in steady state in a dozen or so milliseconds, for example.

In each case, when running the applications with `WAYLAND_DEBUG=1`, I see
```
 -> wp_viewport@30.set_destination(1024, 768)
```
This line is followed by an appreciable pause before the application's window appears. 

I've attached complete debug outputs for one run each of Files, Term, and Edit (captured with `script`). I've added the time index for them the pause occurred for ease.
[edit-debug-long_launch-01.log](https://github.com/pop-os/cosmic-comp/files/14543151/edit-debug-long_launch-01.log) (@ `[2426854.940]`)
[files-debug-long_launch-01.log](https://github.com/pop-os/cosmic-comp/files/14543152/files-debug-long_launch-01.log) (@ `[2395065.880]`)
[term-debug-long_launch-01.log](https://github.com/pop-os/cosmic-comp/files/14543154/term-debug-long_launch-01.log) (@ `[1808805.538]`)

Each of those logs is typical of what I was seeing, and the outputs were otherwise very consistent. I didn't see any significant differences with variations on my display setup. `wp_viewport@30` might become `wp_viewport@28` in a single monitor setup, for example, but the general structure of the behaviour was consistent. 

The frustrating thing is that I'm only seeing this at the moment on a System76 Pang12. When I test a Lenovo X230 laptop, a System76 Pang14, or a  System76 Thelio desktop (with an Intel CPU) I don't see this problem. All three are running the same 6.6.10 kernel, leveraging integrated graphics, have plenty of available memory, and were tested immediately after boot as well just to rule out something specific to how I'm using the Pang12. I also don't think it is likely to be something introduced by specific configuration on my end, as the Pang12 and the X230 are both daily-use long-term installs for me.

**Comment by ryanabx** [comment]
I get about a 3 second delay from click to open with cosmic-files. cosmic-settings doesn't have any delay for me, and cosmic-term has maybe a 1 second delay. Running a hybrid gpu setup with an amd iGPU and amd dGPU.

**Comment by Drakulix** [comment]
Shot in the dark: I have seen similar behaviour with applications calling into dbus-services that are unavailable and waiting for a time out on startup. Could it be possible, that the wayland connection is entirely unrelated to the waiting?

E.g. libcosmic recently introduced portal support for querying dark/light-mode or accent-color, etc. Possibly some of this is causing the slow-down?

**Comment by mmstick** [comment]
Does this also affect cosmic-settings? All but cosmic-settings are using the wgpu renderer, and I've seen the wgpu renderer cause a multi-second delay in initialization. This was affecting cosmic-applets before we reverted back to software rendering.

**Comment by ryanabx** [comment]
> Does this also affect cosmic-settings? All but cosmic-settings are using the wgpu renderer, and I've seen the wgpu renderer cause a multi-second delay in initialization. This was affecting cosmic-applets before we reverted back to software rendering. Some GPUs seem to struggle with initializing wgpu.

I'm not OP, but cosmic-settings was the only one that opened without a delay! You're probably right about it being wgpu

**Comment by Drakulix** [comment]
> Does this also affect cosmic-settings? All but cosmic-settings are using the wgpu renderer, and I've seen the wgpu renderer cause a multi-second delay in initialization. This was affecting cosmic-applets before we reverted back to software rendering. Some GPUs seem to struggle with initializing wgpu.

Oh that would be the case because the nvidia-driver has to actually wake up the GPU for the vulkan-calls the wgpu renderer is using to enumerate GPUs. (Which will happen regardless of us setting the POWER_PREF setting for wgpu. Which is a wgpu bug imo, because it should only open the device send by the wayland protocol.)

**Comment by ryanabx** [comment]
> Which is a wgpu bug imo, because it should only open the device send by the wayland protocol.)

Out of curiosity -- has a bug been filed for it?

**Comment by mmstick** [comment]
I don't see any mentions of this issue

**Comment by git-f0x** [comment]
If it's any help, on the current system that I'm testing COSMIC on (OpenGL 2.1, no Vulkan), there's no delay when opening any app. Since it's likely using software rendering, this might also support that the issue is in wgpu.

**Comment by XV-02** [comment]
I wouldn't expect wgpu to be the culprit on its own, in light of other systems using integrated graphics launching without delay. 

I am seeing delays primarily on a system with a Ryzen 7 6800u with integrated AMD Radeon 680M graphics. 
But on a systems with a Ryzen 7 7840u with AMD Radeon 780M graphics, I don't see this behaviour. 

I haven't tested any hybrid systems explicitly, but don't seem to be seeing it universally even with systems where the graphics device should already be up and running. 

That said, I'll try and build the applications without wgpu and see if the issue persists.

**Comment by ryanabx** [comment]
To clarify, wgpu might have issues with devices with a dGPU, do any of your devices have a dGPU?

**Comment by XV-02** [comment]
Sorry, this fell through on me.

So, this only seems to be a problem for me when I have nvidia-drivers installed on a system which doesn't have nvidia-graphics. 

I do this so that I can test the nvidia-dkms module when updating kernels or nvidia-drivers, as it tends to be a semi-frequent but often quickly fixable blocker. 

Without the drivers installed, or on a system with an actual nvidia graphics device, I don't see the pauses I was seeing. 

Given the oddness of that configuration, I would be totally fine with closing the issue. Alternatively, if this points to something else that needs looking at, let me know.

**Comment by Drakulix** [comment]
> Given the oddness of that configuration, I would be totally fine with closing the issue. Alternatively, if this points to something else that needs looking at, let me know.

At least it should be moved to `libcosmic` or somewhere else appropriate, **if** we want to track this.

This is partly a wgpu bug (I don't think they should enumerate device like they do and rely on the platform - wayland in our case - to tell them which device to use) and partly an nvidia bug, given their driver probably shouldn't search so long for a non-existent device.

But yeah we can close this issue for cosmic-comp, I thknk.

**Comment by mmstick** [comment]
I think it'd be best to file on issue to wgpu directly.

**Comment by ryanabx** [comment]
Agreed. If I could wrap my head around the technical details I'd submit an issue but I don't. The issue does affect me though with an amd dGPU
